contrib/grinder/scripts/utils.py

```diff
@@ -0,0 +1,172 @@
+    properties file determines how many of each type to create based
+    on the "ingest_concurrency" and "query_concurrency" options.
+
+    """
+    thread_type = None
+    server_num = thread_num
+
+    for x in self.types:
+      if server_num < x.num_threads():
+        thread_type = x
+        break
+      else:
+        server_num -= x.num_threads()
+
+    if thread_type == None:
+      raise Exception("Invalid Thread Type")
+
+    return thread_type(thread_num)
+
+
+class AbstractThread(object):
+  #superclass for the various thread types
+  @classmethod
+  def create_metrics(cls, agent_number):
+    raise Exception("Can't create abstract thread")
+
+  @classmethod
+  def num_threads(cls):
+    raise Exception("Can't create abstract thread")
+
+  def make_request(self, logger):
+    raise Exception("Can't create abstract thread")
+
+  def __init__(self, thread_num):
+    # The threads only do so many invocations for each 'report_interval'
+    # position refers to the current position for current interval
+    self.position = 0
+
+    # finish_time is the end time of the interval
+    self.finish_time = int(self.time()) + (default_config['report_interval'] / 1000)
+
+  @classmethod
+  def generate_job_range(cls, total_jobs, total_servers, server_num):
+    """ Determine which subset of the total work the current server is to do.
+
+    The properties file is the same for all the distributed workers and lists the
+    total amount of work to be done for each report interval.  This method allows
+    you to split that work up into the exact subset to be done by the "server_num" worker
+    """
+    jobs_per_server = total_jobs/total_servers
+    remainder = total_jobs % total_servers
+    start_job = jobs_per_server * server_num
+    start_job += min(remainder, server_num)
+    end_job = start_job + jobs_per_server
+    if server_num < remainder:
+      end_job += 1
+    return (start_job, end_job)
+
+  @classmethod
+  def generate_metrics_tenants(cls, num_tenants, metrics_per_tenant, 
+                               agent_number, num_nodes, gen_fn):
+    """ generate the subset of the total metrics to be done by this agent"""
+    tenants_in_shard = range(*cls.generate_job_range(num_tenants, num_nodes, agent_number))
+    metrics = []
+    for y in map(lambda x: gen_fn(x, metrics_per_tenant), tenants_in_shard):
+      metrics += y
+    random.shuffle(metrics)
+    return metrics
+
+  def generate_metric_name(self, metric_id):
+    return default_config['name_fmt'] % metric_id
+
+  def generate_unit(self, tenant_id):
+    unit_number = tenant_id % 6
+    return units_map[unit_number]
+
+  def check_position(self, logger, max_position):
+    """Sleep if finished all work for report interval"""
+    if self.position >= max_position:
+      self.position = 0
+      sleep_time = self.finish_time - int(self.time())
+      self.finish_time += (default_config['report_interval'] / 1000)
```

Comment:
No it won't.  It basically means BF is overloaded and I need to stop and find out where the bottleneck is.